### ISSUE TITLE
CodeAnalyzer false positive for nullability check

### ISSUE THREAD

**Issue by mirasrael** [dotnet/roslyn#49340]
Have these functions which worked before upgrade to MSBuild 16.8 (on 16.7.3 or something like this):
```csharp
[return: MaybeNull] public static TValue GetOrDefault<TKey, TValue>(this IDictionary<TKey, TValue> dict, [AllowNull] TKey key) => key == null ? default : GetOr(dict, key);

[return: MaybeNull] public static TValue GetOr<TKey, TValue>(this IDictionary<TKey, TValue> dict, [AllowNull] TKey key, [AllowNull] TValue def = default) => key != null && dict.TryGetValue(key, out var val) ? val : def;
```
After upgrade compilation fails with:
```
                                                                                                                                                                                                                                                                                  
"W:\projects\Eco-Master\Server\Eco.Shared\Eco.Shared.csproj" (default target) (1) ->                                                                                                                                                                                              
"W:\projects\Eco-Master\Server\Eco.Shared\Eco.Shared.csproj" (Build target) (1:2) ->                                                                                                                                                                                              
(CoreCompile target) ->                                                                                                                                                                                                                                                           
  W:\projects\Eco-Master\Server\Eco.Shared\Utils\ContainerExtensions.cs(1195,169): error CS8620: Argument of type 'IDictionary<TKey, TValue>' cannot be used for parameter 'dict' of type 'IDictionary<TKey, TValue?>' in 'TValue? DictionaryExtensions.GetOr<TKey, TValue?>(IDic 
tionary<TKey, TValue?> dict, TKey key, TValue? def = default(TValue?))' due to differences in the nullability of reference types. [W:\projects\Eco-Master\Server\Eco.Shared\Eco.Shared.csproj]                                                                                    
                                                                                                                                                                                                                                                                                  
                                                                                                                                                                                                                                                                                  
"W:\projects\Eco-Master\Server\Eco.Shared\Eco.Shared.csproj" (default target) (1) ->                                                                                                                                                                                              
"W:\projects\Eco-Master\Server\Eco.Shared\Eco.Shared.csproj" (Build target) (1:3) ->                                                                                                                                                                                              
  W:\projects\Eco-Master\Server\Eco.Shared\Utils\ContainerExtensions.cs(1195,169): error CS8620: Argument of type 'IDictionary<TKey, TValue>' cannot be used for parameter 'dict' of type 'IDictionary<TKey, TValue?>' in 'TValue? DictionaryExtensions.GetOr<TKey, TValue?>(IDic 
tionary<TKey, TValue?> dict, TKey key, TValue? def = default(TValue?))' due to differences in the nullability of reference types. [W:\projects\Eco-Master\Server\Eco.Shared\Eco.Shared.csproj]                                                                                    
```

I can workaround it with (which doesn't seems to add more sense):
```csharp
[return: MaybeNull] public static TValue GetOrDefault<TKey, TValue>(this IDictionary<TKey, TValue> dict, [AllowNull] TKey key) => key == null ? default : GetOr(dict, key, default!);
```

**Comment by huoyaoyuan** [comment]
This may be caused by the new analysis rule of `default` literal in C# 9.0.
I suggest you to change the signature with the new 'nullable unconstrained generic type' of C# 9:
```csharp
public static TValue? GetOrDefault<TKey, TValue>(this IDictionary<TKey, TValue> dict, TKey? key);
public static TValue? GetOr<TKey, TValue>(this IDictionary<TKey, TValue> dict, TKey? key, TValue? def = default);
```

**Comment by mirasrael** [comment]
Yeah, probably. Thanks for advice, but we are not switched to .NET 5 yet and we have explicit language version set in .csproj. So at least it should work with older language versions.
``xml
<LangVersion>8</LangVersion>
```
